File: swift-bridge/Sources/SceneKitBridge/Node.swift
import SceneKit

@_cdecl("scn_node_new")
public func scn_node_new() -> UnsafeMutableRawPointer? {
    scnRetain(SCNNode())
}

@_cdecl("scn_node_new_with_geometry")
public func scn_node_new_with_geometry(_ geometryHandle: UnsafeMutableRawPointer?) -> UnsafeMutableRawPointer? {
    let geometry: SCNGeometry? = scnBorrow(geometryHandle)
    return scnRetain(SCNNode(geometry: geometry))
}

@_cdecl("scn_node_add_child")
public func scn_node_add_child(_ parentHandle: UnsafeMutableRawPointer?, _ childHandle: UnsafeMutableRawPointer?) {
    guard let parent: SCNNode = scnBorrow(parentHandle),
          let child: SCNNode = scnBorrow(childHandle)
    else { return }
    parent.addChildNode(child)
}

@_cdecl("scn_node_remove_from_parent")
public func scn_node_remove_from_parent(_ nodeHandle: UnsafeMutableRawPointer?) {
    guard let node: SCNNode = scnBorrow(nodeHandle) else { return }
    node.removeFromParentNode()
}

@_cdecl("scn_node_copy_name")
public func scn_node_copy_name(_ nodeHandle: UnsafeMutableRawPointer?) -> UnsafeMutablePointer<CChar>? {
    guard let node: SCNNode = scnBorrow(nodeHandle) else { return nil }
    return scnDup(node.name)
}

@_cdecl("scn_node_set_name")
public func scn_node_set_name(_ nodeHandle: UnsafeMutableRawPointer?, _ name: UnsafePointer<CChar>?) {
    guard let node: SCNNode = scnBorrow(nodeHandle) else { return }
    node.name = name.map(String.init(cString:))
}

@_cdecl("scn_node_get_transform")
public func scn_node_get_transform(_ nodeHandle: UnsafeMutableRawPointer?, _ outMatrix: UnsafeMutableRawPointer?) -> Bool {
    guard let node: SCNNode = scnBorrow(nodeHandle) else { return false }
    return scnWriteMatrix4(node.transform, out: outMatrix)
}

@_cdecl("scn_node_set_transform")
public func scn_node_set_transform(_ nodeHandle: UnsafeMutableRawPointer?, _ matrixHandle: UnsafeMutableRawPointer?) {
    guard let node: SCNNode = scnBorrow(nodeHandle), let matrix = scnReadMatrix4(matrixHandle) else { return }
    node.transform = matrix
}

@_cdecl("scn_node_get_position")
public func scn_node_get_position(_ nodeHandle: UnsafeMutableRawPointer?, _ outVector: UnsafeMutableRawPointer?) -> Bool {
    guard let node: SCNNode = scnBorrow(nodeHandle) else { return false }
    return scnWriteVector3(node.position, out: outVector)
}

@_cdecl("scn_node_set_position")
public func scn_node_set_position(_ nodeHandle: UnsafeMutableRawPointer?, _ vectorHandle: UnsafeMutableRawPointer?) {
    guard let node: SCNNode = scnBorrow(nodeHandle), let vector = scnReadVector3(vectorHandle) else { return }
    node.position = vector
}

@_cdecl("scn_node_get_rotation")
public func scn_node_get_rotation(_ nodeHandle: UnsafeMutableRawPointer?, _ outVector: UnsafeMutableRawPointer?) -> Bool {
    guard let node: SCNNode = scnBorrow(nodeHandle) else { return false }
    return scnWriteVector4(node.rotation, out: outVector)
}

@_cdecl("scn_node_set_rotation")
public func scn_node_set_rotation(_ nodeHandle: UnsafeMutableRawPointer?, _ vectorHandle: UnsafeMutableRawPointer?) {
    guard let node: SCNNode = scnBorrow(nodeHandle), let vector = scnReadVector4(vectorHandle) else { return }
    node.rotation = vector
}

@_cdecl("scn_node_get_scale")
public func scn_node_get_scale(_ nodeHandle: UnsafeMutableRawPointer?, _ outVector: UnsafeMutableRawPointer?) -> Bool {
    guard let node: SCNNode = scnBorrow(nodeHandle) else { return false }
    return scnWriteVector3(node.scale, out: outVector)
}

@_cdecl("scn_node_set_scale")
public func scn_node_set_scale(_ nodeHandle: UnsafeMutableRawPointer?, _ vectorHandle: UnsafeMutableRawPointer?) {
    guard let node: SCNNode = scnBorrow(nodeHandle), let vector = scnReadVector3(vectorHandle) else { return }
    node.scale = vector
}

@_cdecl("scn_node_get_euler_angles")
public func scn_node_get_euler_angles(_ nodeHandle: UnsafeMutableRawPointer?, _ outVector: UnsafeMutableRawPointer?) -> Bool {
    guard let node: SCNNode = scnBorrow(nodeHandle) else { return false }
    return scnWriteVector3(node.eulerAngles, out: outVector)
}

@_cdecl("scn_node_set_euler_angles")
public func scn_node_set_euler_angles(_ nodeHandle: UnsafeMutableRawPointer?, _ vectorHandle: UnsafeMutableRawPointer?) {
    guard let node: SCNNode = scnBorrow(nodeHandle), let vector = scnReadVector3(vectorHandle) else { return }
    node.eulerAngles = vector
}

@_cdecl("scn_node_get_pivot")
public func scn_node_get_pivot(_ nodeHandle: UnsafeMutableRawPointer?, _ outMatrix: UnsafeMutableRawPointer?) -> Bool {
    guard let node: SCNNode = scnBorrow(nodeHandle) else { return false }
    return scnWriteMatrix4(node.pivot, out: outMatrix)
}

@_cdecl("scn_node_set_pivot")
public func scn_node_set_pivot(_ nodeHandle: UnsafeMutableRawPointer?, _ matrixHandle: UnsafeMutableRawPointer?) {
    guard let node: SCNNode = scnBorrow(nodeHandle), let matrix = scnReadMatrix4(matrixHandle) else { return }
    node.pivot = matrix
}

@_cdecl("scn_node_get_hidden")
public func scn_node_get_hidden(_ nodeHandle: UnsafeMutableRawPointer?) -> Bool {
    guard let node: SCNNode = scnBorrow(nodeHandle) else { return false }
    return node.isHidden
}

@_cdecl("scn_node_set_hidden")
public func scn_node_set_hidden(_ nodeHandle: UnsafeMutableRawPointer?, _ hidden: Bool) {
    guard let node: SCNNode = scnBorrow(nodeHandle) else { return }
    node.isHidden = hidden
}

@_cdecl("scn_node_get_geometry")
public func scn_node_get_geometry(_ nodeHandle: UnsafeMutableRawPointer?) -> UnsafeMutableRawPointer? {
    guard let node: SCNNode = scnBorrow(nodeHandle), let geometry = node.geometry else { return nil }
    return scnRetain(geometry)
}

@_cdecl("scn_node_set_geometry")
public func scn_node_set_geometry(_ nodeHandle: UnsafeMutableRawPointer?, _ geometryHandle: UnsafeMutableRawPointer?) {
    guard let node: SCNNode = scnBorrow(nodeHandle) else { return }
    let geometry: SCNGeometry? = scnBorrow(geometryHandle)
    node.geometry = geometry
}

@_cdecl("scn_node_get_light")
public func scn_node_get_light(_ nodeHandle: UnsafeMutableRawPointer?) -> UnsafeMutableRawPointer? {
    guard let node: SCNNode = scnBorrow(nodeHandle), let light = node.light else { return nil }
    return scnRetain(light)
}

@_cdecl("scn_node_set_light")
public func scn_node_set_light(_ nodeHandle: UnsafeMutableRawPointer?, _ lightHandle: UnsafeMutableRawPointer?) {
    guard let node: SCNNode = scnBorrow(nodeHandle) else { return }
    let light: SCNLight? = scnBorrow(lightHandle)
    node.light = light
}

@_cdecl("scn_node_get_camera")
public func scn_node_get_camera(_ nodeHandle: UnsafeMutableRawPointer?) -> UnsafeMutableRawPointer? {
    guard let node: SCNNode = scnBorrow(nodeHandle), let camera = node.camera else { return nil }
    return scnRetain(camera)
}

@_cdecl("scn_node_set_camera")
public func scn_node_set_camera(_ nodeHandle: UnsafeMutableRawPointer?, _ cameraHandle: UnsafeMutableRawPointer?) {
    guard let node: SCNNode = scnBorrow(nodeHandle) else { return }
    let camera: SCNCamera? = scnBorrow(cameraHandle)
    node.camera = camera
}

@_cdecl("scn_node_get_physics_body")
public func scn_node_get_physics_body(_ nodeHandle: UnsafeMutableRawPointer?) -> UnsafeMutableRawPointer? {
    guard let node: SCNNode = scnBorrow(nodeHandle), let body = node.physicsBody else { return nil }
    return scnRetain(body)
}

@_cdecl("scn_node_set_physics_body")
public func scn_node_set_physics_body(_ nodeHandle: UnsafeMutableRawPointer?, _ physicsBodyHandle: UnsafeMutableRawPointer?) {
    guard let node: SCNNode = scnBorrow(nodeHandle) else { return }
    let physicsBody: SCNPhysicsBody? = scnBorrow(physicsBodyHandle)
    node.physicsBody = physicsBody
}

@_cdecl("scn_node_run_action")
public func scn_node_run_action(_ nodeHandle: UnsafeMutableRawPointer?, _ actionHandle: UnsafeMutableRawPointer?) {
    guard let node: SCNNode = scnBorrow(nodeHandle),
          let action: SCNAction = scnBorrow(actionHandle)
    else { return }
    node.runAction(action)
}

@_cdecl("scn_node_child_nodes")
public func scn_node_child_nodes(_ nodeHandle: UnsafeMutableRawPointer?) -> UnsafeMutableRawPointer? {
    guard let node: SCNNode = scnBorrow(nodeHandle) else { return nil }
    return scnRetain(node.childNodes as NSArray)
}

@_cdecl("scn_node_get_parent")
public func scn_node_get_parent(_ nodeHandle: UnsafeMutableRawPointer?) -> UnsafeMutableRawPointer? {
    guard let node: SCNNode = scnBorrow(nodeHandle), let parent = node.parent else { return nil }
    return scnRetain(parent)
}

@_cdecl("scn_node_child_node_with_name")
public func scn_node_child_node_with_name(
    _ nodeHandle: UnsafeMutableRawPointer?,
    _ name: UnsafePointer<CChar>?,
    _ recursively: Bool
) -> UnsafeMutableRawPointer? {
    guard let node: SCNNode = scnBorrow(nodeHandle), let name,
          let child = node.childNode(withName: String(cString: name), recursively: recursively)
    else { return nil }
    return scnRetain(child)
}

@_cdecl("scn_node_clone")
public func scn_node_clone(_ nodeHandle: UnsafeMutableRawPointer?) -> UnsafeMutableRawPointer? {
    guard let node: SCNNode = scnBorrow(nodeHandle) else { return nil }
    let clone = node.clone()
    scnAdoptRendererDelegates(from: node, to: clone)
    return scnRetain(clone)
}

@_cdecl("scn_node_get_world_transform")
public func scn_node_get_world_transform(_ nodeHandle: UnsafeMutableRawPointer?, _ outMatrix: UnsafeMutableRawPointer?) -> Bool {
    guard let node: SCNNode = scnBorrow(nodeHandle) else { return false }
    return scnWriteMatrix4(node.worldTransform, out: outMatrix)
}

@_cdecl("scn_node_set_world_transform")
public func scn_node_set_world_transform(_ nodeHandle: UnsafeMutableRawPointer?, _ matrixHandle: UnsafeMutableRawPointer?) {
    guard let node: SCNNode = scnBorrow(nodeHandle), let matrix = scnReadMatrix4(matrixHandle) else { return }
    node.setWorldTransform(matrix)
}

@_cdecl("scn_node_get_opacity")
public func scn_node_get_opacity(_ nodeHandle: UnsafeMutableRawPointer?) -> Double {
    guard let node: SCNNode = scnBorrow(nodeHandle) else { return 1 }
    return Double(node.opacity)
}

@_cdecl("scn_node_set_opacity")
public func scn_node_set_opacity(_ nodeHandle: UnsafeMutableRawPointer?, _ opacity: Double) {
    guard let node: SCNNode = scnBorrow(nodeHandle) else { return }
    node.opacity = CGFloat(opacity)
}

@_cdecl("scn_node_get_category_bit_mask")
public func scn_node_get_category_bit_mask(_ nodeHandle: UnsafeMutableRawPointer?) -> UInt {
    guard let node: SCNNode = scnBorrow(nodeHandle) else { return 0 }
    return UInt(bitPattern: node.categoryBitMask)
}

@_cdecl("scn_node_set_category_bit_mask")
public func scn_node_set_category_bit_mask(_ nodeHandle: UnsafeMutableRawPointer?, _ mask: UInt) {
    guard let node: SCNNode = scnBorrow(nodeHandle) else { return }
    node.categoryBitMask = Int(bitPattern: mask)
}

File: src/node.rs
use core::ptr;

use crate::action::Action;
use crate::camera::Camera;
use crate::error::take_string;
use crate::ffi;
use crate::geometry::Geometry;
use crate::light::Light;
use crate::math::{Matrix4, Vector3, Vector4};
use crate::physics::PhysicsBody;
use crate::private::{cstring_from_str, handle_type};

handle_type!(Node);

impl Node {
    /// Creates a wrapped `SCNNode` instance.
    #[must_use]
    pub fn new() -> Option<Self> {
        unsafe { Self::from_raw(ffi::scn_node_new()) }
    }

    /// Mirrors `SCNNode.withGeometry`.
    #[must_use]
    pub fn with_geometry(geometry: Option<&Geometry>) -> Option<Self> {
        unsafe {
            Self::from_raw(ffi::scn_node_new_with_geometry(
                geometry.map_or(ptr::null_mut(), Geometry::as_ptr),
            ))
        }
    }

    /// Mirrors `SCNNode.addChildNode`.
    pub fn add_child_node(&self, child: &Self) {
        unsafe { ffi::scn_node_add_child(self.ptr, child.ptr) };
    }

    /// Mirrors `SCNNode.removeFromParent`.
    pub fn remove_from_parent(&self) {
        unsafe { ffi::scn_node_remove_from_parent(self.ptr) };
    }

    /// Mirrors `SCNNode.name`.
    #[must_use]
    pub fn name(&self) -> Option<String> {
        unsafe { take_string(ffi::scn_node_copy_name(self.ptr)) }
    }

    /// Sets the `SCNNode.name` member.
    pub fn set_name(&self, name: &str) {
        if let Some(name) = cstring_from_str(name) {
            unsafe { ffi::scn_node_set_name(self.ptr, name.as_ptr()) };
        }
    }

    /// Mirrors `SCNNode.transform`.
    #[must_use]
    pub fn transform(&self) -> Matrix4 {
        let mut matrix = Matrix4::default();
        let _ = unsafe { ffi::scn_node_get_transform(self.ptr, matrix.as_mut_ptr().cast()) };
        matrix
    }

    /// Sets the `SCNNode.transform` member.
    pub fn set_transform(&self, transform: Matrix4) {
        unsafe { ffi::scn_node_set_transform(self.ptr, transform.as_ptr().cast_mut().cast()) };
    }

    /// Mirrors `SCNNode.position`.
    #[must_use]
    pub fn position(&self) -> Vector3 {
        let mut value = Vector3::default();
        let _ = unsafe { ffi::scn_node_get_position(self.ptr, value.as_mut_ptr().cast()) };
        value
    }

    /// Sets the `SCNNode.position` member.
    pub fn set_position(&self, position: Vector3) {
        unsafe { ffi::scn_node_set_position(self.ptr, position.as_ptr().cast_mut().cast()) };
    }

    /// Mirrors `SCNNode.rotation`.
    #[must_use]
    pub fn rotation(&self) -> Vector4 {
        let mut value = Vector4::default();
        let _ = unsafe { ffi::scn_node_get_rotation(self.ptr, value.as_mut_ptr().cast()) };
        value
    }

    /// Sets the `SCNNode.rotation` member.
    pub fn set_rotation(&self, rotation: Vector4) {
        unsafe { ffi::scn_node_set_rotation(self.ptr, rotation.as_ptr().cast_mut().cast()) };
    }

    /// Mirrors `SCNNode.scale`.
    #[must_use]
    pub fn scale(&self) -> Vector3 {
        let mut value = Vector3::default();
        let _ = unsafe { ffi::scn_node_get_scale(self.ptr, value.as_mut_ptr().cast()) };
        value
    }

    /// Sets the `SCNNode.scale` member.
    pub fn set_scale(&self, scale: Vector3) {
        unsafe { ffi::scn_node_set_scale(self.ptr, scale.as_ptr().cast_mut().cast()) };
    }

    /// Mirrors `SCNNode.eulerAngles`.
    #[must_use]
    pub fn euler_angles(&self) -> Vector3 {
        let mut value = Vector3::default();
        let _ = unsafe { ffi::scn_node_get_euler_angles(self.ptr, value.as_mut_ptr().cast()) };
        value
    }

    /// Sets the `SCNNode.eulerAngles` member.
    pub fn set_euler_angles(&self, euler_angles: Vector3) {
        unsafe {
            ffi::scn_node_set_euler_angles(self.ptr, euler_angles.as_ptr().cast_mut().cast());
        };
    }

    /// Mirrors `SCNNode.pivot`.
    #[must_use]
    pub fn pivot(&self) -> Matrix4 {
        let mut matrix = Matrix4::default();
        let _ = unsafe { ffi::scn_node_get_pivot(self.ptr, matrix.as_mut_ptr().cast()) };
        matrix
    }

    /// Sets the `SCNNode.pivot` member.
    pub fn set_pivot(&self, pivot: Matrix4) {
        unsafe { ffi::scn_node_set_pivot(self.ptr, pivot.as_ptr().cast_mut().cast()) };
    }

    /// Returns the `SCNNode.isHidden` value.
    #[must_use]
    pub fn is_hidden(&self) -> bool {
        unsafe { ffi::scn_node_get_hidden(self.ptr) }
    }

    /// Sets the `SCNNode.hidden` member.
    pub fn set_hidden(&self, hidden: bool) {
        unsafe { ffi::scn_node_set_hidden(self.ptr, hidden) };
    }

    /// Mirrors `SCNNode.geometry`.
    #[must_use]
    pub fn geometry(&self) -> Option<Geometry> {
        unsafe { Geometry::from_raw(ffi::scn_node_get_geometry(self.ptr)) }
    }

    /// Sets the `SCNNode.geometry` member.
    pub fn set_geometry(&self, geometry: Option<&Geometry>) {
        unsafe {
            ffi::scn_node_set_geometry(
                self.ptr,
                geometry.map_or(ptr::null_mut(), Geometry::as_ptr),
            );
        };
    }

    /// Mirrors `SCNNode.light`.
    #[must_use]
    pub fn light(&self) -> Option<Light> {
        unsafe { Light::from_raw(ffi::scn_node_get_light(self.ptr)) }
    }

    /// Sets the `SCNNode.light` member.
    pub fn set_light(&self, light: Option<&Light>) {
        unsafe { ffi::scn_node_set_light(self.ptr, light.map_or(ptr::null_mut(), Light::as_ptr)) };
    }

    /// Mirrors `SCNNode.camera`.
    #[must_use]
    pub fn camera(&self) -> Option<Camera> {
        unsafe { Camera::from_raw(ffi::scn_node_get_camera(self.ptr)) }
    }

    /// Sets the `SCNNode.camera` member.
    pub fn set_camera(&self, camera: Option<&Camera>) {
        unsafe {
            ffi::scn_node_set_camera(self.ptr, camera.map_or(ptr::null_mut(), Camera::as_ptr));
        };
    }

    /// Mirrors `SCNNode.physicsBody`.
    #[must_use]
    pub fn physics_body(&self) -> Option<PhysicsBody> {
        unsafe { PhysicsBody::from_raw(ffi::scn_node_get_physics_body(self.ptr)) }
    }

    /// Sets the `SCNNode.physicsBody` member.
    pub fn set_physics_body(&self, physics_body: Option<&PhysicsBody>) {
        unsafe {
            ffi::scn_node_set_physics_body(
                self.ptr,
                physics_body.map_or(ptr::null_mut(), PhysicsBody::as_ptr),
            );
        };
    }

    /// Mirrors `SCNNode.runAction`.
    pub fn run_action(&self, action: &Action) {
        unsafe { ffi::scn_node_run_action(self.ptr, action.as_ptr()) };
    }
}
